programs/ziti-edge-tunnel/windows-scripts.c

```diff
@@ -0,0 +1,356 @@
+/*
+ Copyright 2019-2021 NetFoundry Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#include <ziti/ziti_log.h>
+#include <model/dtos.h>
+#include "ziti/ziti_tunnel.h"
+#include "windows/windows-scripts.h"
+
+#define MAX_BUCKET_SIZE 512
+
+static char* const namespace_template = "%s@{n='%s';}";
+static char* const exe_name = "ziti-tunnel";
+
+struct hostname_s {
+    char *hostname;
+    LIST_ENTRY(hostname_s) _next;
+};
+typedef LIST_HEAD(hostname_list_s, hostname_s) hostname_list_t;
+
+static void exit_cb(uv_process_t* process,
+                    int64_t exit_status,
+                    int term_signal) {
+    ZITI_LOG(TRACE, "Process exited with status %d, signal %d", exit_status, term_signal);
+    uv_close((uv_handle_t*)process, (uv_close_cb) free);
+}
+
+static bool exec_process(uv_loop_t *ziti_loop, const char* program, const char* args[]) {
+    uv_process_t* process = calloc(1, sizeof(uv_process_t));
+    uv_process_options_t options = {0};
+    uv_stdio_container_t stdio[3];
+    options.file = program;
+    options.args = args;
+    options.exit_cb = exit_cb;
+
+    options.stdio = stdio;
+    options.stdio_count = 3;
+    options.stdio[0].flags = UV_IGNORE;
+    options.stdio[1].flags = UV_INHERIT_FD;
+    options.stdio[1].data.fd = 1;
+    options.stdio[2].flags = UV_INHERIT_FD;
+    options.stdio[2].data.fd = 2;
+
+    int r = uv_spawn(ziti_loop, process, &options);
+    if (r != 0) {
+        ZITI_LOG(ERROR, "Could not execute the command due to %s", uv_err_name(r));
+        return false;
+    }
+    uv_unref((uv_handle_t*) process);
+    return true;
+}
+
+static char* exec_process_fetch_result(char* program) {
+    FILE *fp;
+    char path[BUFFER_SIZE];
+
+    fp = popen(program, "r");
+    if (fp == NULL) {
+        ZITI_LOG(ERROR,"Failed to run command %s", strerror(errno));
+        return NULL;
+    }
+    char* result = calloc(MAXBUFFERLEN, sizeof(char));
+
+    while (fgets(path, sizeof(path), fp) != NULL) {
+        strcat(result, path);
+    }
+
+    pclose(fp);
+
+    return result;
+}
+
+void chunked_add_nrpt_rules(uv_loop_t *ziti_loop, hostname_list_t *hostnames, char* tun_ip) {
+    char* script = calloc(MAX_POWERSHELL_SCRIPT_LEN, sizeof(char));
+    size_t buf_len = sprintf(script, "$Namespaces = @(");
```

Comment:
We need to guarantee that `script` is not overrun. I think a good way to do that is to replace `sprintf` with `snprintf`.